xds/src/main/java/io/grpc/xds/Filter.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Message;
+import io.grpc.ClientInterceptor;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.ServerInterceptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * Defines the parsing functionality of an HTTP filter. A Filter may optionally implement either
+ * {@link ClientInterceptorBuilder} or {@link ServerInterceptorBuilder} or both, indicating it is
+ * capable of working on the client side or server side or both, respectively.
+ */
+interface Filter {
+
+  /**
+   * The proto message types supported by this filter. A filter will be registered by each of its
+   * supported message types.
+   */
+  String[] typeUrls();
+
+  /**
+   * Parses the top-level filter config from raw proto message. The message may be either a {@link
+   * com.google.protobuf.Any} or a {@link com.google.protobuf.Struct}.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage);
+
+  /**
+   * Parses the per-filter override filter config from raw proto message. The message may be either
+   * a {@link com.google.protobuf.Any} or a {@link com.google.protobuf.Struct}.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfigOverride(Message rawProtoMessage);
+
+  /** Represents an opaque data structure holding configuration for a filter. */
+  interface FilterConfig {
+    String typeUrl();
+  }
+
+  /** Uses the FilterConfigs produced above to produce an HTTP filter interceptor for clients. */
+  interface ClientInterceptorBuilder {
+    @Nullable
+    ClientInterceptor buildClientInterceptor(
```

Comment:
This interface is redundant. It looks you don't need to passing both the top-level filter config and the override config(s). It should be as simple as `buildClientInterceptor(FilterConfig config, ...)`. The caller should just need to pass in the single filter config: the most fine-grained one (WeightedCluster > Route > VirtualHost > HttpConnectionManager). The existing interface is half-baked: you let the implementation decide whether to use `config` or `overrideConfig`, but let the caller decide which filter config to be passed in as the `overrideConfig`.

Given that, it can be further changed to have this method on `FilterConfig` interface:

```java
interfacce FilterConfig {
  String typeUrl();

  ClientInterceptor getClientInterceptor(PickSubchannelArgs args, ScheduledExecutorService scheduler);
}
```
as the same type of FilterConfig always produces the same kind of ClientInterceptor.

Reply:
The spec says:
"Note that, for a given request, the filter's configuration will come from the top-level filter list in the HttpConnectionManager config and the most specific override. "

Both [c-core](https://github.com/grpc/grpc/blob/master/src/core/ext/xds/xds_http_filters.h#L101-L103) and [go](https://github.com/grpc/grpc-go/blob/master/xds/internal/httpfilter/httpfilter.go#L63) APIs are passing both the top-level and override configs.

Reply:
Fault filter uses just override config if it is provided, and that's just a special case, and for general HttpFilters that's not the case.

Reply:
I am commenting from interface/class design's perspective. I don't think direct copying method signatures from other languages is appropriate, and IMO such an interface looks very awkward.

If you want a general purpose interface for merging the top-level config and its overrides (could be up to 3 levels of overrides), the method should take in a list of overriding configs, not just the most specific one. If it is well-defined that merging configs is just letting the most specific one take over, the method only needs to take in a single config argument where as the caller always passes in the most specific config. And the method of returning a ClientInterceptor can just be associated with the FilterConfig instance, as the implementation of converting a certain type of FilterConfig to a ClientInterceptor is specific to each type of config itself.

Anyway, I can just let you pass if you just want to do exactly the same thing as C-core and Go.

Reply:
The interface is for passing a top-level config and a merged override config (the 3 level of overrides merged as the finalized override), and in general, the top-level config and the merged override config are of different types, and both of them will be needed to generate ClientInterceptor.  The interface is for a general _HttpFilter_.  For the Fault Filter, the top-level and override configs are of the same type, and if override config is provided, the top-level will not be used, but that's a special implementation.